pkg/providers/ssm/provider.go

```diff
@@ -60,3 +74,54 @@ func (p *DefaultProvider) Get(ctx context.Context, parameter string) (string, er
 	log.FromContext(ctx).WithValues("parameter", parameter, "value", result.Parameter.Value).Info("discovered ssm parameter")
 	return lo.FromPtr(result.Parameter.Value), nil
 }
+
+// Invalidate asyncronously invalidates entries in the underlying SSM cache. Invalidation is scheduled up to 30 minutes
+// in the future to avoid thundering herds caused by globally correlated invalidations. Any subsequent invalidations of
+// the same parameter made before the initial invalidation has been processed are no-ops.
+func (p *DefaultProvider) Invalidate(ctx context.Context, parameters ...string) {
+	p.Lock()
+	defer p.Unlock()
+
+	nonPending := map[string]string{}
+	for _, parameter := range parameters {
+		if p.pendingInvalidations.Has(parameter) {
+			continue
+		}
+		currentValue, ok := p.cache.Get(parameter)
+		if !ok {
+			continue
+		}
+		nonPending[parameter] = currentValue.(string)
+	}
+	if len(nonPending) == 0 {
+		return
+	}
+	p.pendingInvalidations.Insert(lo.Keys(nonPending)...)
```

Comment:
If you're just adding the keys to this set, why do you even care about the values of nonPending? Can nonPending be a set as well?

Reply:
We track the current value in `nonPending` that way we only invalidate the cache entry if it's still the same when the invalidation request is processed. Not strictly necessary, but this dedupes SSM calls. This is getting dropped anyway though.